GH-2615 and GH-2997- Added an updated guide for Windows Server 2016 (#3141)
mattermost/docs: source/deployment/sso-saml-adfs-msws2016.rst

Configure SAML with Microsoft ADFS using Microsoft Windows Server 2016
======================================================================

This document provides steps to configure SAML 2.0 with Microsoft ADFS for Mattermost and Microsoft Windows Server 2016.

.. contents::
  :backlinks: top
  :local:

.. include:: sso-saml-before-you-begin.rst

> ## Prerequisites

Reviewer: ```suggestion
Prerequisites
-------------------------
```

Configure SAML with Microsoft ADFS using Microsoft Windows Server 2016
======================================================================

> The following process provides steps to configure SAML 2.0 with Microsoft ADFS for Mattermost and Microsoft Windows Server 2016.

Reviewer 2: ```suggestion
This document provides steps to configure SAML 2.0 with Microsoft ADFS for Mattermost and Microsoft Windows Server 2016.
```

.. contents::
  :backlinks: top
  :local:

.. include:: sso-saml-before-you-begin.rst

> The following are basic requirements to use ADFS for Mattermost:

Reviewer 2: ```suggestion
## Prerequisites
```

>  - An Active Directory instance where all users have a specified email and username attributes. For Mattermost servers running 3.3 and earlier, users must also have their first name and last name attributes specified.

Reviewer 2: ```suggestion
 - An Active Directory instance where all users have specified email and username attributes. For Mattermost servers running 3.3 and earlier, users must also have their first name and last name attributes specified.
```

>  - A Microsoft Server running. The screenshots used in this guide are from Microsoft Server 2012R2, but similar steps should work for other versions.

Reviewer 2: ```suggestion
 - A running Microsoft Server. The screenshots used in this guide are from Microsoft Server 2012R2, but similar steps should work for other versions.
```

 - An SSL certificate to sign your ADFS login page.
 - ADFS installed on your Microsoft Server. You can find a detailed guide for deploying and configuring ADFS in `this article <https://msdn.microsoft.com/en-us/library/gg188612.aspx>`__.

> On your ADFS installation, note down the value of the **SAML 2.0/W-Federation URL** in ADFS Endpoints section, also known as the **SAML SSO URL Endpoint** in this guide. If you chose the defaults for the installation, this will be ``/adfs/ls/``.

Reviewer 2: Would it be possible to include the path to the ADFS Endpoints section? So, something like: 
```suggestion
On your ADFS installation, navigate to "Menu name" > ADFS Endpoints and note down the value of the **SAML 2.0/W-Federation URL**/**SAML SSO URL Endpoint**. If you chose the defaults for the installation, this will be ``/adfs/ls/``.
```

Add a Relying Party Trust
-------------------------

> 1. In ADFS management snap-in, go to **AD FS > Relying Party Trusts** and click **Add Relying Party Trust** on the right sidebar. As an alternative right-click **Relying Party Trusts** and choose **Add Relying Party Trust** from the context menu.

Reviewer 2: ```suggestion
1. Open the ADFS management snap-in and select **ADFS > Relying Party Trusts > Add Relying Party Trust** on the right sidebar. You can also right-click **Relying Party Trusts** and choose **Add Relying Party Trust** from the context menu.
```

	.. image:: ../../source/images/SSO-SAML-ADFS_add-new-relying-party-trust_000.png

> 2. A configuration wizard for adding a new relying party trust opens. In the **Welcome** screen, choose **Claims aware** and click **Start**.

Reviewer 2: ```suggestion
2. In the **Welcome** screen of the configuration wizard choose **Claims aware** and select **Start**.
```

	.. image:: ../../source/images/SSO-SAML-ADFS_add-new-relying-party-trust_001.png

> 3. In the **Select Data Source** screen, select the option **Enter data about the relying party manually**.

Reviewer 2: ```suggestion
3. In the **Select Data Source** screen, choose **Enter data about the relying party manually**.
```

	.. image:: ../../source/images/SSO-SAML-ADFS_add-new-relying-party-trust_002.png

> 4. In the **Specify Display Name** screen, enter a **Display Name** to recognize the trust, such as ``Mattermost``, and add any notes you want to make.

Reviewer 2: ```suggestion
4. In the **Specify Display Name** screen enter a **Display Name** (e.g., ``Mattermost``). You can add optional notes.
```

	.. image:: ../../source/images/SSO-SAML-ADFS_add-new-relying-party-trust_003.png

5. In the **Configure Certificate** screen, leave the certificate settings at their default values.

	.. image:: ../../source/images/SSO-SAML-ADFS_add-new-relying-party-trust_004.png

> However, if you would like to set up encryption for your SAML connection, click the **Browse** button and upload your Service Provider Public Certificate.

Reviewer 2: ```suggestion
If you would like to set up encryption for your SAML connection, select **Browse** and upload your Service Provider Public Certificate.
```

	.. image:: ../../source/images/SSO-SAML-ADFS_add-new-relying-party-trust_005.png

> 6. In the **Configure URL** screen, select the option **Enable Support for the SAML 2.0 WebSSO protocol** and enter the **SAML 2.0 SSO service URL**, which is of the form ``https://<your-mattermost-url>/login/sso/saml`` where ``https://<your-mattermost-url>`` should typically match the `Mattermost Site URL <https://docs.mattermost.com/administration/config-settings.html#site-url>`__.

Reviewer 2: ```suggestion
6. In the **Configure URL** screen, select **Enable Support for the SAML 2.0 WebSSO protocol** and enter the **SAML 2.0 SSO service URL** in the following format:``https://<your-mattermost-url>/login/sso/saml`` where ``https://<your-mattermost-url>`` should typically match the `Mattermost Site URL <https://docs.mattermost.com/administration/config-settings.html#site-url>`__.
```

	.. image:: ../../source/images/SSO-SAML-ADFS_add-new-relying-party-trust_006.png

> 7. In the **Configure Identifiers** screen, enter the **Relying party trust identifier** (also known as the **Identity Provider Issuer URL**) of the form ``https://<your-idp-url>/adfs/services/trust`` and click **Add**.

Reviewer 2: ```suggestion
7. In the **Configure Identifiers** screen, enter the **Relying party trust identifier** (also known as the **Identity Provider Issuer URL**) in the following format ``https://<your-idp-url>/adfs/services/trust`` and choose **Add**.
```

	.. image:: ../../source/images/SSO-SAML-ADFS_add-new-relying-party-trust_007.png

	.. image:: ../../source/images/SSO-SAML-ADFS_add-new-relying-party-trust_008.png

> Add your **SAML 2.0 SSO service URL** from above the same way.

Reviewer 2: Is this added to the "Configure Identifiers" menu as a separate entry, or is it in the same process?

Author: The process is identical to the one above using a different URL, resulting in two separate entries in the "Configure Identifiers" step.

> 8. In the **Choose Access Control Policy** screen select the access control policy suitable for your environment. This guide assumes the default values **Permit everyone** and an unchecked box are kept.

Reviewer 2: ```suggestion
8. In the **Choose Access Control Policy** screen select the access control policy suitable for your environment. This guide assumes the default values **Permit everyone** and an unchecked box.
```

	.. image:: ../../source/images/SSO-SAML-ADFS_add-new-relying-party-trust_009.png

9. In the **Ready to Add Trust** screen, you can review your settings.

	.. image:: ../../source/images/SSO-SAML-ADFS_add-new-relying-party-trust_010.png

> 10. In the **Finish** screen, select the option **Configure claims issuance policy for this application**, and click **Close**. 

Reviewer 2: ```suggestion
10. In the **Finish** screen select **Configure claims issuance policy for this application** and then **Close**. 
```

	.. image:: ../../source/images/SSO-SAML-ADFS_add-new-relying-party-trust_011.png

> 	You will now exit the configuration wizard and a **Claim Rules** editor opens.

Reviewer 2: I think this can be removed as the previous step indicated the wizard would be closed and the next section refers to the Claim Rules creation process.

Author: agreed and removed

Create Claim Rules
------------------

> 1. In the **Issuance Transform Rules** tab of the **Claim Rules** editor, click the **Add Rule…** button.

Reviewer 2: ```suggestion
1. In the **Issuance Transform Rules** tab of the **Claim Rules** editor, choose **Add Rule…**.
```

	.. image:: ../../source/images/SSO-SAML-ADFS_create-claim-rules_001.png

> 	This action opens an **Add Transform Claim Rule Wizard**.

Reviewer 2: I think this line can be removed.

Author: agreed and removed

2. In the **Choose Rule Type** screen, select **Send LDAP Attributes as Claims** from the drop-down menu, then click **Next**.

	.. image:: ../../source/images/SSO-SAML-ADFS_create-claim-rules_002.png

3. In the **Configure Claim Rule** screen, enter a **Claim Rule Name** of your choice, select **Active Directory** as the **Attribute Store** and add the following mapping:
  - From the LDAP Attribute column, **select** ``E-Mail-Addresses``. From the Outgoing Claim Type, **type** ``Email``
  - From the LDAP Attribute column, **select** ``Given-Name``. From the Outgoing Claim Type, **type** ``FirstName``
  - From the LDAP Attribute column, **select** ``Surname``. From the Outgoing Claim Type, **type** ``LastName``
  - From the LDAP Attribute column, **select** ``SAM-Account-Name``. From the Outgoing Claim Type, **type** ``Username``

For Mattermost 3.4 and later, the *FirstName* and *LastName* attributes are optional.

> Then, click **Finish** to add the rule.

Reviewer 2: ```suggestion
Select **Finish** to add the rule.
```

> Note that the entries in the **Outgoing Claim Type** column can be chosen to be something else. They can contain dashes but no spaces. Note that they will be used to map the corresponding fields in Mattermost later.

Reviewer 2: ```suggestion
The entries in the **Outgoing Claim Type** column can be modified and the entries may contain dashes but no spaces. They are used to map the corresponding fields in Mattermost.
```

	.. image:: ../../source/images/SSO-SAML-ADFS_create-claim-rules_003.png

> 4. Create another new rule by clicking the **Add Rule** button.

Reviewer 2: ```suggestion
4. Select **Add Rule** to create another new rule.
```

5. In the **Choose Rule Type** screen, select **Transform an Incoming Claim** from the drop-down menu, then click **Next**.

	.. image:: ../../source/images/SSO-SAML-ADFS_create-claim-rules_004.png

6. In the **Configure Claim Rule** screen, enter a **Claim Rule Name** of your choice, then
  - Select *Name ID* for the **Incoming claim type**
  - Select *Unspecified* for the **Incoming name ID format**
  - Select *E-Mail Address* for the **Outgoing claim type**

> Moreover, select the **Pass through all claim values** option. Then click **Finish**.

Reviewer 2: ```suggestion
Select the **Pass through all claim values** option and click **Finish**.
```

	.. image:: ../../source/images/SSO-SAML-ADFS_create-claim-rules_005.png

7. Click **Finish** to create the claim rule, then **OK** to finish creating rules.

8. Open Windows PowerShell as an administrator and run the following command:

  ``Set-ADFSRelyingPartyTrust -TargetName <display-name> -SamlResponseSignature "MessageAndAssertion"``

> where <display-name> is the name you specified in step 6. In our example it would be ``mattermost``.

Reviewer 2: ```suggestion
where *<display-name>* is the name you specified in step 6. In our example it would be ``mattermost``.
```
Could you confirm that "step 6" refers to step 6 immediately preceding this? Or is it another step 6?

Author: Thanks for finding this, it was actually step 4 of Add a Relying Party Trust. Corrected it.

This action will add the signature to SAML messages, making verification successful.

Export Identity Provider Certificate
-------------------------------------

Next, we export the identity provider certificate, which will be later uploaded to Mattermost to finish SAML configuration.

> 1. In ADFS management snap-in, go to **AD FS > Service > Certificates** and double click on the certificate under **Token-signing**. You may alternatively right-click the field, then click **View Certificate** in the context menu.

Reviewer 2: ```suggestion
1. Open the ADFS management snap-in, select **AD FS > Service > Certificates** and double click on the certificate under **Token-signing**. You can also right-click the field, then click **View Certificate** in the context menu.
```

	.. image:: ../../source/images/SSO-SAML-ADFS_export-id-provider-cert_001.png

> 2. In the **Certificate** screen, go to the **Details** tab and click **Copy to File**, then **OK**.

Reviewer 2: ```suggestion
2. In the **Certificate** screen open the the **Details** tab, select **Copy to File** and then **OK**.
```

	.. image:: ../../source/images/SSO-SAML-ADFS_export-id-provider-cert_003.png

> 	This opens the **Certificate Export Wizard**.

Reviewer 2: I think this line can be removed as the screenshot is self-explanatory.

Author: agreed and removed

3. In the **Certificate Export Wizard** screen, click **Next**.

	.. image:: ../../source/images/SSO-SAML-ADFS_export-id-provider-cert_004.png

	Then, select the option **Base-64 encoded X.509 (.CER)** and click **Next** again.

	.. image:: ../../source/images/SSO-SAML-ADFS_export-id-provider-cert_005.png

4. In the **Certificate Export Wizard** screen, click **Browse** to specify the location you want the Identity Provider Certificate to be exported, and specify the file name.

	.. image:: ../../source/images/SSO-SAML-ADFS_export-id-provider-cert_006.png

5. Click **Save**. In the **Certificate Export Wizard** screen, verify the file path is correct, and click **Next**.

6. In the **Completing the Certificate Export Wizard**, click **Finish**, then **OK** to confirm the export was successful.

	.. image:: ../../source/images/SSO-SAML-ADFS_export-id-provider-cert_007.png

You’re now about to finish configuring SAML for Mattermost!

> Configure SAML sign-in for Mattermost

Reviewer 2: ```suggestion
Configure SAML Sign-in for Mattermost
```

--------------------------------------

1. Start Mattermost server and sign into Mattermost as a System Administrator. Go to **System Console > Authentication > SAML**.

>   - Set *Enable Login With SAML 2.0* to *true*.

Reviewer 2: ```suggestion
  - Set **Enable Login With SAML 2.0** to ``true``.
```

>   - Set *Enable Synchronizing SAML Accounts With AD/LDAP* to suit your environment.

Reviewer 2: ```suggestion
  - Set **Enable Synchronizing SAML Accounts With AD/LDAP** to suit your environment.
```

>   - Set *Override SAML bind data with AD/LDAP information* to suit your environment.

Reviewer 2: ```suggestion
  - Set **Override SAML bind data with AD/LDAP information** to suit your environment.
```

>   - For *SAML SSO URL* use ``SAML 2.0/W-Federation URL ADFS Endpoint`` you copied earlier.

Reviewer 2: ```suggestion
  - For **SAML SSO URL** use the ``SAML 2.0/W-Federation URL ADFS Endpoint`` you copied at the beginning of the process.
```

>   - For Identity Provider Issuer URL* use ``Relying party trust identifier`` from ADFS you specified earlier.

Reviewer 2: ```suggestion
  - For **Identity Provider Issuer URL** use the ``Relying party trust identifier`` from ADFS.
```

>   - For *Identity Provider Public Certificate* use ``X.509 Public Certificate`` you downloaded earlier.

Reviewer 2: ```suggestion
  - For **Identity Provider Public Certificate** use  the``X.509 Public Certificate``. 
```

	.. image:: ../../source/images/SSO-SAML-ADFS_configure-saml_001.png

2. Configure Mattermost to verify the signature. 

>   - Set *Verify Signature* to *true*.

Reviewer 2: ```suggestion
  - Set **Verify Signature** to ``true``.
```

>   - For *Service Provider Login URL* use the ``SAML 2.0 SSO service URL`` you specified in ADFS earlier.

Reviewer 2: ```suggestion
  - For **Service Provider Login URL** use the ``SAML 2.0 SSO service URL`` you specified in ADFS.
```

	.. image:: ../../source/images/SSO-SAML-ADFS_configure-saml_002.png

3. Enable encryption.

>   - Set *Enable Encryption* to *true*.

Reviewer 2: ```suggestion
  - Set **Enable Encryption** to ``true``.
```

>   - For *Service Provider Private Key* use the Service Provider Private Key you generated at the beginning.

Reviewer 2: ```suggestion
  - For **Service Provider Private Key** use the Service Provider Private Key generated at the start of this process.
```

>   - For *Service Provider Public Certificate* use the Service Provider Public Certificate you generated at the beginning.

Reviewer 2: ```suggestion
  - For **Service Provider Public Certificate** use the Service Provider Public Certificate you generated at the start of this process.
```

>   - Set *Sign Request* to suit your environment.

Reviewer 2: ```suggestion
  - Set **Sign Request** to suit your environment.
```

	.. image:: ../../source/images/SSO-SAML-ADFS_configure-saml_003.png

4. Set attributes for the SAML Assertions, which will be used to update user information in Mattermost. Attributes for email and username are required and should match the values you entered in ADFS earlier. See :ref:`documentation on SAML configuration settings <saml-enterprise>` for more detail.

For Mattermost servers running 3.3 and earlier, the first name and last name attributes are also required fields.

	.. image:: ../../source/images/SSO-SAML-ADFS_configure-saml_004.png

5. Click **Save**.

> 6. (Optional) If you configured First Name Attribute and Last Name Attribute, go to **System Console > General > Users and Teams** in prior versions or **System Console** > **Site Configuration** > **Users and Teams** in versions after 5.12 and set **Teammate Name Display** to *Show first and last name*. This is recommended for a better user experience.

Reviewer 2: ```suggestion
6. (Optional) If you configured First Name Attribute and Last Name Attribute, go to **System Console > Site Configuration > Users and Teams** (or **System Console > General > Users and Teams** in versions prior to 5.12) and set **Teammate Name Display** to **Show first and last name**. This is recommended for a better user experience.
```
